README.md
# Luzifer / ed-fast-travel

`ed-fast-travel` is a small helper utility to plan long-distance travels inside [Elite: Dangerous](https://www.elitedangerous.com).

## Facts

- It uses the nightly database dump from [EDDB](https://www.eddb.io/)
- The database is cached locally in `~/.local/share/ed-fast-travel`
- It works better when calculating stops in 100 Ly or more distance on the route
- You can plot courses with more than 25000 Ly distance in ~3m

## Difficulties

- Maybe the system you want to travel to is not yet known in EDSM/EDDB  
> As of today 0.001313% of the galaxy has been discovered on EDSM, it will take 132,444 years, 4 months and 19 days to discover it entirely.
- Calculations with ~25 Ly stop distance might work but maybe they contain jumps >30 Ly distance as of above reason
- Calculations with <20 Ly distance might even run into an endless loop as of above reason
- Inside is no real route-planning engine but only a simple vector calculation to find stops on a linear axis

## Usage

You can download pre-compiled binaries on [Github releases](https://github.com/Luzifer/ed-fast-travel/releases/latest) for your system.

### Web-Interface

Starting with version `v0.7.0` `ed-fast-travel` does contain a GUI to be used through your browser. To use that one just start the command without any arguments (you may want to see the parameters for more adjustments but the defaults will work for most people) and it will listen on port 3000 on your local machine. To access the GUI just access `http://localhost:3000/` and you can start using it:

![GUI Demo](https://d3vv6lp55qjaqc.cloudfront.net/items/0L1N3R3r2W1w051U1k2B/Screen%20Recording%202016-09-13%20at%2001.51%20PM.gif)

On the first start the initialization will require a few moments as the database needs to be downloaded and loaded into the application.

For all parameters available please execute the utility with `--help` parameter.

### CLI / Text Route-Planning

To get a route in plain text format you need to execute the tool using the console / cmd prompt:

[![asciicast](https://asciinema.org/a/7ea5fd8hexx9wy38bcge3er1j.png)](https://asciinema.org/a/7ea5fd8hexx9wy38bcge3er1j?t=12)

As you can see you only need to know from where you're starting and where you're going. The number (500 in above case) is the distance the stops should be calcualated. I used 500 Ly as that's a good distance for the ingame route engine to calculate a more detailed route.

### Updates

`ed-fast-travel` contains an own updater and will print an information if an update is available:

```
2016/09/11 21:46:09 An update to ed-fast-travel is available. Run ./ed-fast-travel --self-update to update.
```

### Database Updates

To update the local database with a fresh nightly dump from EDSM just add the flag `--update` when executing the utility:

```bash
# ./ed-fast-travel --update
2016/09/11 17:21:56 No local EDSM dump found or update forced, fetching dump...
Usage: ed-fast-travel <start system> <target system> <distance between nav points>

Example: ed-fast-travel 'Sol' 'Dryooe Prou GL-Y d369' 500
  This will calculate stops on your route from Sol to Dryooe Prou GL-Y d369 every 500Ly
```

The usage explanation appears because no parameters for routing are passed in that execution but the database now is refreshed. Please be nice to the servers and refresh only if you are sure you need to get new data and that there is new data. Refreshing multiple times a day does not give you any advantages as the database is only updated once a day.

## Build from source

- To build this utility from source you need to have a working go1.7 environment.
- Execute `go get github.com/Luzifer/ed-fast-travel` and you will get the `ed-fast-travel` binary inside your `$GOPATH/bin/` directory


History.md
# 0.15.1 / 2016-12-14

  * Remove local publish-script
  * Download vault2env from Github
  * Build database on travis
  * Move gzip compression to unix level

# 0.15.0 / 2016-10-29

  * Migrate to EDDB v5 API

# 0.14.0 / 2016-10-19

  * Drop windows/386 support
  * Several minor fixes in README
  * Do not break if no tag is present
  * Create new releases as drafts

# 0.13.0 / 2016-10-12

  * Change update mechanism to Github releases

# 0.12.1 / 2016-10-12

  * Add builder configuration for pushes to Github releases
  * Exit after generate database

# 0.12.0 / 2016-10-02

  * Switch to EDDB data, display permit and scoopables

# 0.11.0 / 2016-10-02

  * Improve CLI logging
  * Implement backtracking to calculate routes with low jump ranges
  * Introduce bucket database for huge speed increase
  * Restructure systems database for faster access

# 0.10.0 / 2016-09-13

  * Add database update to frontend

# 0.9.0 / 2016-09-13

  * Add german translation
  * Add i18n framework for web frontend

# 0.8.2 / 2016-09-13

  * Fix: File format was missing argument

# 0.8.1 / 2016-09-13

  * Fix: Respect distance when caching

# 0.8.0 / 2016-09-13

  * Switch to binary storage format

# 0.7.0 / 2016-09-13

  * Add route caching
  * Implement web interface

# 0.6.0 / 2016-09-11

  * Disable update check when using version built from source

# 0.5.1 / 2016-09-11

  * Fix self-update

# 0.5.0 / 2016-09-11

  * Add autoupdater

# 0.4.0 / 2016-09-11

  * Use own strong compressed cache for dump data

# 0.3.0 / 2016-09-11

  * Add flag to silence log ouput

# 0.2.1 / 2016-09-11

  * Add instructions for manual build

# 0.2.0 / 2016-09-11

  * Disable color output on windows by default

# 0.1.2 / 2016-09-11

  * Fix gobuilder builds

# 0.1.1 / 2016-09-11

  * Vendor dependencies
  * Add asciicast instead of text demo

# 0.1.0 / 2016-09-11

  * Initial version


main.go
package main

//go:generate go-bindata -pkg $GOPACKAGE -o assets.go assets/ i18n/

import (
	"compress/gzip"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"path"
	"sync"
	"time"

	"github.com/Luzifer/rconfig"
	"github.com/fatih/color"
	homedir "github.com/mitchellh/go-homedir"
)

const (
	autoUpdateRepo        = "Luzifer/ed-fast-travel"
	autoUpdateLabel       = "master"
	readableDumpName      = "dump_v3.bin"
	compressedDumpName    = "dump_v3.bin.gz"
	edsmDumpURL           = "http://assets.luzifer.io/" + compressedDumpName
	originalSystemDumpURL = "https://eddb.io/archive/v5/systems.csv"
	originalBodyDumpURL   = "https://eddb.io/archive/v5/bodies.jsonl"
)

var (
	cfg = struct {
		CacheURI               string        `flag:"cache" default:"file://~/.local/share/ed-fast-travel/cache" description:"Cache to use for route caching"`
		CacheTime              time.Duration `flag:"cache-time" default:"1h" description:"How long to cache routes"`
		Color                  bool          `flag:"color" vardefault:"color" description:"Use color for output"`
		DisableSoftwareControl bool          `flag:"disable-software-control" default:"false" description:"Do not let web-users control update / shutdown"`
		EDSMDumpPath           string        `flag:"data-path" default:"~/.local/share/ed-fast-travel" description:"Path to store EDSM data"`
		Generate               bool          `flag:"generate-database" default:"false" description:"Instead of downloading the database generate it from original dump"`
		Listen                 string        `flag:"listen" default:":3000" description:"IP/Port to listen on when starting in web mode"`
		SelfUpdate             bool          `flag:"self-update" default:"false" description:"Update the tool to the latest version"`
		Silent                 bool          `flag:"silent,s" default:"false" description:"Suppress every message except the flight plan"`
		UpdateData             bool          `flag:"update" default:"false" description:"Fetch latest dump from EDSM"`
		VersionAndExit         bool          `flag:"version" default:"false" description:"Prints current version and exits"`
		WebRouteTimeout        time.Duration `flag:"web-route-timeout" default:"10m" description:"Timout for route calculations requested via web interface"`
		WebRouteStopMin        float64       `flag:"web-route-stop-min" default:"100" description:"Min distance between stops"`
	}{}

	version = "dev"

	starSystems     *starSystemDatabase
	starSystemsLock sync.RWMutex
	cache           routeCache
)

func init() {
	rconfig.SetVariableDefaults(defaultSettings)
	if err := rconfig.Parse(&cfg); err != nil {
		log.Fatalf("Unable to parse commandline options: %s", err)
	}

	color.NoColor = !cfg.Color

	if cfg.VersionAndExit {
		fmt.Printf("ed-fast-travel %s\n", version)
		os.Exit(0)
	}

	if cfg.SelfUpdate {
		selfUpdate()
		os.Exit(1)
	}

	var err error
	cfg.EDSMDumpPath, err = homedir.Expand(cfg.EDSMDumpPath)
	if err != nil {
		log.Fatalf("Could not expand data-path: %s", err)
	}

	cache, err = getCache(cfg.CacheURI)
	if err != nil {
		log.Fatalf("Could not find specified caching method")
	}
}

func verboseLog(format string, args ...interface{}) {
	if !cfg.Silent {
		log.Printf(format, args...)
	}
}

func printHelp() {
	fmt.Println("Usage for web interface:  ed-fast-travel")
	fmt.Println("Usage for CLI calculation: ed-fast-travel <start system> <target system> <distance between nav points>\n")
	fmt.Println("Example: ed-fast-travel 'Sol' 'Dryooe Prou GL-Y d369' 500\n  This will calculate stops on your route from Sol to Dryooe Prou GL-Y d369 every 500Ly")
}

func main() {
	checkUpdates()

	if cfg.Generate {
		if err := generateGOBDatabase(); err != nil {
			log.Fatalf("Could not generate database: %s", err)
		}
		os.Exit(0)
	}

	if _, err := os.Stat(path.Join(cfg.EDSMDumpPath, readableDumpName)); err != nil || cfg.UpdateData {
		if err := refreshEDSMData(); err != nil {
			log.Fatalf("Unable to refresh EDSM data: %s", err)
		}
	}

	// Load database
	verboseLog("Loading database...")
	var err error
	starSystems, err = loadStarSystems()
	if err != nil {
		log.Fatalf("Could not load star systems from dump: %s", err)
	}

	switch len(rconfig.Args()) {
	case 4:
		doCLICalculation()
	case 1:
		startWebService()
	default:
		printHelp()
		os.Exit(1)
	}
}

func refreshEDSMData() error {
	verboseLog("No local EDSM dump found or update forced, fetching dump...")
	if err := os.MkdirAll(cfg.EDSMDumpPath, 0755); err != nil {
		return err
	}

	if _, err := os.Stat(path.Join(cfg.EDSMDumpPath, "etag.txt")); err == nil {
		d, err := ioutil.ReadFile(path.Join(cfg.EDSMDumpPath, "etag.txt"))
		if err != nil {
			return err
		}
		eTag := string(d)

		resp, err := http.Head(edsmDumpURL)
		if err != nil {
			return err
		}
		defer resp.Body.Close()

		if resp.Header.Get("ETag") == eTag {
			return nil
		}
	}

	dump, err := os.Create(path.Join(cfg.EDSMDumpPath, readableDumpName))
	if err != nil {
		return err
	}
	defer dump.Close()

	resp, err := http.Get(edsmDumpURL)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	cdata, err := gzip.NewReader(resp.Body)
	if err != nil {
		return err
	}

	_, err = io.Copy(dump, cdata)
	if err != nil {
		return err
	}

	if err := ioutil.WriteFile(path.Join(cfg.EDSMDumpPath, "etag.txt"), []byte(resp.Header.Get("ETag")), 0644); err != nil {
		return err
	}

	if !cfg.SelfUpdate {
		go func() {
			starSystemsLock.Lock()
			defer starSystemsLock.Unlock()

			// Update was triggered from web interface
			starSystems, err = loadStarSystems()
			if err != nil {
				log.Fatalf("Could not load star systems from dump: %s", err)
				return
			}

			log.Printf("Database reloaded as requested from web interface.")
		}()
	}

	return nil
}
